### PR TITLE
refactoring pilosa/server

### PR DESCRIPTION
**Pull request by jaffee**
## Overview

trying to separate internal an external concerns in pilosa.Server - it should
handle Cluster, Holder, etc. while pilosa/server handles things with external
deps - e.g. Logger, Stats, Handler, etc. Using functional options in
pilosa.Server now.

Fixes #

## Pull request checklist

- [ ] I have read the [contributing guide](https://github.com/pilosa/pilosa/blob/master/CONTRIBUTING.md).
- [ ] I have agreed to the [Contributor License Agreement](https://cla-assistant.io/pilosa/pilosa).
- [ ] I have updated the [documentation](https://github.com/pilosa/pilosa/tree/master/docs).
- [ ] I have resolved any merge conflicts.
- [ ] I have included tests that cover my changes.
- [ ] All new and existing tests pass.

## Code review checklist
This is the checklist that the reviewer will follow while reviewing your pull request. You do not need to do anything with this checklist, but be aware of what the reviewer will be looking for.

- [ ] Ensure that any changes to external docs have been included in this pull request.
- [ ] If the changes require that minor/major versions need to be updated, tag the PR appropriately.
- [ ] Ensure the new code is [properly commented](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and follows [Idiomatic Go](https://dmitri.shuralyov.com/idiomatic-go).
- [ ] Check that tests have been written and that they cover the new functionality.
- [ ] Run tests and ensure they pass.
- [ ] Build and run the code, performing any applicable integration testing.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -55,64 +55,161 @@ func OptServerLogger(l Logger) ServerOption {
...
+	return func(s *Server) error {
+		s.NewAttrStore = af
+		s.Holder.NewAttrStore = af
+		return nil
+	}
+}
+
+func OptServerAntiEntropyInterval(interval time.Duration) ServerOption {
+	return func(s *Server) error {
+		s.AntiEntropyInterval = interval
```
Comment (travisturner) [inline server.go]: I didn't check all of these functional options, but we can un-export `AntiEntropyInterval` now unless there's a reason to leave it exported.

File: server/server.go [hunk 2]
```diff
@@ -125,77 +119,73 @@ func (m *Command) Run(args ...string) (err error) {
...
 	}
 
-	m.Server.Logger.Printf("Listening as %s\n", m.Server.URI)
+	m.logger.Printf("Listening as %s\n", m.Server.URI)
+
 	return nil
 }
 
+// Wait waits for the server to be closed or interrupted.
+func (m *Command) Wait() error {
```
Comment (travisturner) [inline server/server.go]: the `Run`/`Wait` terminology is odd to me, but I understand what's going on. I also don't have a better suggestion right now.

File: server/server.go [hunk 3]
```diff
@@ -125,77 +119,73 @@ func (m *Command) Run(args ...string) (err error) {
...
+		return errors.Wrap(m.Close(), "closing command")
+	case <-m.Done:
+		m.logger.Printf("Server closed externally")
+		return nil
+	}
+}
+
 // SetupLogger sets up the logger based on the configuration.
-func (m *Command) SetupLogger() error {
+func (m *Command) SetupLogger() (pilosa.Logger, error) {
```
Comment (travisturner) [inline server/server.go]: does this need to be exported?

File: server/server.go [hunk 4]
```diff
@@ -318,3 +354,24 @@ func NewStatsClient(name string, host string) (pilosa.StatsClient, error) {
 		return nil, errors.Errorf("'%v' not a valid stats client, choose from [expvar, statsd, none].")
 	}
 }
+
+// OpenListener opens a listener for the Server.
```
Comment (travisturner) [inline server/server.go]: fix comment to match function name

**Comment by jaffee** [comment]
@travisturner I think your concerns are addressed now - Todd and I have plans to continue this work in the morning, but it might be better to get this portion merged first.

**Pull request merged**